### PR TITLE
refactor: Fix circular dependency with storage callbacks

### PR DESCRIPTION
**Pull request by mykola-vrmchk**
### Motivation / Background

This Pull Request ...

### CODE REVIEW

**Comment by mykola-vrmchk** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dxos/dxos/6530?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6530** <a href="https://app.graphite.dev/github/pr/dxos/dxos/6530?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6527** <a href="https://app.graphite.dev/github/pr/dxos/dxos/6527?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mykola-vrmchk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by yarolegovich — APPROVED** [review]

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 1]
```diff
@@ -159,6 +167,31 @@ export class AutomergeHost {
...
+      return;
+    }
+    const doc = handle.docSync();
+    if (!doc) {
+      return;
+    }
+
+    const lastAvailableHash = getHeads(doc);
+
+    const getDocumentObjects = (doc: SpaceDoc): string[] => Object.keys(doc.objects ?? {});
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:183]: іs this function only used on the next line?

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 2]
```diff
@@ -159,6 +167,31 @@ export class AutomergeHost {
     return this._repo;
   }
 
+  private async _beforeSave({ path, batch }: BeforeSaveParams) {
+    const getDocumentIdFromPath = (path: StorageKey): DocumentId => path[0] as DocumentId;
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:171]: іs this function only used on the next line?

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 3]
```diff
@@ -159,6 +167,31 @@ export class AutomergeHost {
...
+    const getDocumentObjects = (doc: SpaceDoc): string[] => Object.keys(doc.objects ?? {});
+
+    const objectIds = getDocumentObjects(doc);
+    const encodedIds = objectIds.map((objectId) => idCodec.encode({ documentId: handle.documentId, objectId }));
+    const idToLastHash = new Map(encodedIds.map((id) => [id, lastAvailableHash]));
+    this._indexMetadataStore.markDirty(idToLastHash, batch);
+  }
+
+  private async _afterSave() {
+    this._indexMetadataStore.afterMarkDirty();
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:192]: In general function names should represent the actions they perform (as oposed to callbacks):

```suggestion
    this._indexMetadataStore.notifyMarkedDirty();
```

+ add jsdoc to that method

**Review by dmaretskyi — DISMISSED** [review]

**Pull request merged**